### PR TITLE
Performance improvements

### PR DESCRIPTION
**Pull request by dnadales**
- Fix a stack overflow error caused by a lazy `Map.fold` being used.
- Compute the key hashes only once.
- Make sure all the benchmark data is evaluated before running the benchmarks.

### CODE REVIEW

File: formal-spec/participants-vs-elapsed_time.dat [hunk 1]
```diff
@@ -1,6 +1,6 @@
-1.00E+02	4.01E-06
-1.00E+03	4.23E-05
+1.00E+02	4.04E-06
+1.00E+03	4.22E-05
 1.00E+04	4.53E-04
-1.00E+05	6.70E-03
-1.00E+06	7.50E-02
-1.00E+07	2.92E+00
+1.00E+05	7.55E-03
+1.00E+06	8.37E-02
```
Comment (dnadales) [inline formal-spec/participants-vs-elapsed_time.dat]: @nkarag does this mean that the processing time went up by ~100 milliseconds, for 1M participants, but went down by 1 second for 10M participants?

File: formal-spec/measurements.tex [hunk 2]
```diff
@@ -262,7 +262,7 @@ \subsection{Processing Time and Memory Consumption}
   \label{fig:eltime-vs-participants}
 \end{figure}
 
-In Figure \ref{fig:eltime-vs-participants}, we see that the elapsed 
+In Figure \ref{fig:eltime-vs-participants}, we see that the processing
```
Comment (dnadales) [inline formal-spec/measurements.tex]: I would also add a comment on whether this memory consumption seems reasonable. I'd say it does:

```
Haskell maps use 6 words (a word takes 8 bytes) per node. In general there are twice as many nodes as values. Each value requires two words. And storing a hash requires a word for the tag, plus the size of the hash. Then, for storing p hashes we need:

  p * 8 * (6 + 6 +  2 + 1 + 4)

So if p = 1M we need 152 MB to store the hashes, so the resulting 195 MB used to tally the votes of 1M  participants is within the memory usage range we'd expect.
```

Should we add something like that?

File: formal-spec/participants-vs-elapsed_time.dat (hunk 1 above)
Comment (nkarag) [reply formal-spec/participants-vs-elapsed_time.dat]: Yeah. That is what it seems. These are the numbers!

File: formal-spec/measurements.tex (hunk 2 above)
Comment (nkarag) [reply formal-spec/measurements.tex]: Sure why not.

**Pull request merged**